### ISSUE TITLE
Reusing matchers as asymmetric equality testers

### ISSUE THREAD

**Issue by triskweline** [jasmine/jasmine#2039]
### Your question

I often find myself implementing the same assertions twice:

1. as a custom matcher, to match individual values.
2. as an asymmetric equality tester, for partial matching in `toHaveBeenCalledWith()` and `toEqual()`.

Is it somehow possible to implement the logic only once, and then use it for both cases?

### Example code that demonstrates the problem

For example, this code tests if a DOM element matches a given CSS selector:

```JavaScript
// (1) As a custom matcher
beforeEach(function () {
  jasmine.addMatchers({
    toMatchSelector(util) {
      return {
        compare(actualElement, expectedSelector) {
          return { pass: actualElement.matches(expectedSelector) }
        }
      }
    }
  })
})

// (2) As an asymmetric equality tester
function elementMatchingSelector(expectedSelector) {
  return {
    asymmetricMatch(actualElement) {
      return actualElement.matches(expectedSelector)
    },

    jasmineToString() {
      return `<elementMatchingSelector: ${expectedSelector}>`
    }
  }
}
```

### jasmine-core version

5.2.0

### Node.js or browser version

Chrome 127

### Operating System

Ubuntu 20.04 LTS

**Comment by sgravrock** [comment]
The most straightforward thing is probably to extract the logic into a regular function and call it from both places. But you could also have the matcher use the custom equality tester directly:

```
  jasmine.addMatchers({
    toMatchSelector(util) {
      return {
        compare(actualElement, expectedSelector) {
          return { pass: elementMatchingSelector(expectedSelector).asymmetricMatch(actualElement) }
        }
      }
    }
```

Or via `equals`, although that's probably unnecessarily complicated in most cases:

```
  jasmine.addMatchers({
    toMatchSelector(util) {
      return {
        compare(actualElement, expectedSelector) {
          return { pass: util.equals(actualElement, elementMatchingSelector(expectedSelector)) }
        }
      }
    }
```

**Comment by triskweline** [comment]
I guess if you're too lazy to define a matcher, one more option would be to use the asymmetric matcher with `toEqual()`:

```js
expect(element).toEqual(elementMatchingSelector('#foo'))
```

... albeit at the cost of some readability.

Maybe there's an opportunity for an API that creates both matcher and asymmetric tester from a single object, but that's out of scope for this question. Thanks @sgravrock !
